De-prioritise Dimname and DimnameList in python overload resolution

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50903 Add correction parameter to std/var
* **#51350 De-prioritise Dimname and DimnameList in python overload resolution**

`None` being a valid `Dimname` is awkward for optional `dim` arguments, as found
on NumPy's reduction functions like `std` and `var`. In these cases `dim=None`
should mean an all-reduction, but instead you get an error
"Please look up dimensions by name".

I've also had to fix `FunctionParameter::check` to actually check the first
element of `INT_LIST` arguments and reject non-int types. Otherwise, the dim
names end up calling the `int[]` overload and fail.

Differential Revision: [D26756208](https://our.internmc.facebook.com/intern/diff/D26756208)

peterbell10 committed:
De-prioritise Dimname and DimnameList in python overload resolution

`None` being a valid `Dimname` is awkward for optional `dim` arguments, as found
on NumPy's reduction functions like `std` and `var`. In this cases `dim=None`
should mean an all-reduction, but instead you get an error
"Please look up dimensions by name".

I've also had to fix `FunctionParameter::check` to actually check the first
element of `INT_LIST` arguments and reject non-int types. Otherwise, the dim
names end up calling the `int[]` overload and fail.

[ghstack-poisoned]

peterbell10 committed:
Update on "De-prioritise Dimname and DimnameList in python overload resolution"


`None` being a valid `Dimname` is awkward for optional `dim` arguments, as found
on NumPy's reduction functions like `std` and `var`. In these cases `dim=None`
should mean an all-reduction, but instead you get an error
"Please look up dimensions by name".

I've also had to fix `FunctionParameter::check` to actually check the first
element of `INT_LIST` arguments and reject non-int types. Otherwise, the dim
names end up calling the `int[]` overload and fail.

[ghstack-poisoned]

peterbell10 committed:
Update on "De-prioritise Dimname and DimnameList in python overload resolution"


`None` being a valid `Dimname` is awkward for optional `dim` arguments, as found
on NumPy's reduction functions like `std` and `var`. In these cases `dim=None`
should mean an all-reduction, but instead you get an error
"Please look up dimensions by name".

I've also had to fix `FunctionParameter::check` to actually check the first
element of `INT_LIST` arguments and reject non-int types. Otherwise, the dim
names end up calling the `int[]` overload and fail.

[ghstack-poisoned]

peterbell10 committed:
Update on "De-prioritise Dimname and DimnameList in python overload resolution"


`None` being a valid `Dimname` is awkward for optional `dim` arguments, as found
on NumPy's reduction functions like `std` and `var`. In these cases `dim=None`
should mean an all-reduction, but instead you get an error
"Please look up dimensions by name".

I've also had to fix `FunctionParameter::check` to actually check the first
element of `INT_LIST` arguments and reject non-int types. Otherwise, the dim
names end up calling the `int[]` overload and fail.

[ghstack-poisoned]

mruberry commented:
@bhosmer I think you're the correct reviewer for this change?

bhosmer commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -472,7 +472,14 @@ auto FunctionParameter::check(PyObject* obj, std::vector<py::handle> &overloaded
+            PySequence_GetItem(obj, 0));
+        // NOTE: JIT tracer allows arbitrary scalar tensors to act as ints
+        // in an intlist argument Even float or complex scalars
+        return (THPVariable_Check(item.ptr()) || THPUtils_checkIndex(item.ptr()));
```
Be good to factor this out now that it's nontrivial, a la `is_float_or_complex_list()` just below

bhosmer reviewed (commented):
Given the awkwardness of `Dimname` scooping up `None` values, this seems like a reasonable solution for the eager binding layer. I think we'll need to do something similar for the JIT though. cc @eellison for confirmation - Elias, do we support Dimname overloads on the JIT side?

bhosmer reviewed (commented):
Given the awkwardness of `Dimname` scooping up `None` values, this seems like a reasonable solution for the eager binding layer. I think we'll need to do something similar for the JIT though. cc @eellison for confirmation - Elias, do we support Dimname overloads on the JIT side?

peterbell10 committed:
Update on "De-prioritise Dimname and DimnameList in python overload resolution"


`None` being a valid `Dimname` is awkward for optional `dim` arguments, as found
on NumPy's reduction functions like `std` and `var`. In these cases `dim=None`
should mean an all-reduction, but instead you get an error
"Please look up dimensions by name".

I've also had to fix `FunctionParameter::check` to actually check the first
element of `INT_LIST` arguments and reject non-int types. Otherwise, the dim
names end up calling the `int[]` overload and fail.

[ghstack-poisoned]

peterbell10 committed:
Update on "De-prioritise Dimname and DimnameList in python overload resolution"


`None` being a valid `Dimname` is awkward for optional `dim` arguments, as found
on NumPy's reduction functions like `std` and `var`. In these cases `dim=None`
should mean an all-reduction, but instead you get an error
"Please look up dimensions by name".

I've also had to fix `FunctionParameter::check` to actually check the first
element of `INT_LIST` arguments and reject non-int types. Otherwise, the dim
names end up calling the `int[]` overload and fail.

[ghstack-poisoned]

peterbell10 committed:
Update on "De-prioritise Dimname and DimnameList in python overload resolution"


`None` being a valid `Dimname` is awkward for optional `dim` arguments, as found
on NumPy's reduction functions like `std` and `var`. In these cases `dim=None`
should mean an all-reduction, but instead you get an error
"Please look up dimensions by name".

I've also had to fix `FunctionParameter::check` to actually check the first
element of `INT_LIST` arguments and reject non-int types. Otherwise, the dim
names end up calling the `int[]` overload and fail.

[ghstack-poisoned]

eellison commented:
@bhosmer i dont think dimname is supported, the schemas with dimname dont get attempted to be matched to in JIT

bhosmer commented:
> @bhosmer i dont think dimname is supported, the schemas with dimname dont get attempted to be matched to in JIT. it probably wouldnt be too hard to add just no one has done it :/

Ah got it. Well, I don't think observing this overload ordering will add any difficulty once someone does, at least. Thanks for the info!

bhosmer reviewed (approved):
Looks good, thanks!

peterbell10 committed:
Update on "De-prioritise Dimname and DimnameList in python overload resolution"


`None` being a valid `Dimname` is awkward for optional `dim` arguments, as found
on NumPy's reduction functions like `std` and `var`. In these cases `dim=None`
should mean an all-reduction, but instead you get an error
"Please look up dimensions by name".

I've also had to fix `FunctionParameter::check` to actually check the first
element of `INT_LIST` arguments and reject non-int types. Otherwise, the dim
names end up calling the `int[]` overload and fail.

[ghstack-poisoned]

peterbell10 committed:
Update on "De-prioritise Dimname and DimnameList in python overload resolution"


`None` being a valid `Dimname` is awkward for optional `dim` arguments, as found
on NumPy's reduction functions like `std` and `var`. In these cases `dim=None`
should mean an all-reduction, but instead you get an error
"Please look up dimensions by name".

I've also had to fix `FunctionParameter::check` to actually check the first
element of `INT_LIST` arguments and reject non-int types. Otherwise, the dim
names end up calling the `int[]` overload and fail.

[ghstack-poisoned]

peterbell10 committed:
Update on "De-prioritise Dimname and DimnameList in python overload resolution"


`None` being a valid `Dimname` is awkward for optional `dim` arguments, as found
on NumPy's reduction functions like `std` and `var`. In these cases `dim=None`
should mean an all-reduction, but instead you get an error
"Please look up dimensions by name".

I've also had to fix `FunctionParameter::check` to actually check the first
element of `INT_LIST` arguments and reject non-int types. Otherwise, the dim
names end up calling the `int[]` overload and fail.

[ghstack-poisoned]
